### DISCUSSION TITLE
how to use Ml with dart stack

### DISCUSSION THREAD

**Discussion by rectrick** [serverpod/serverpod#1017] [Q&A] [answered]
how to use Ml with dart stack with out python or django ? i need use all ml with serverpod and flutter can we get it done?

**Comment by Isakdl** [comment, marked as answer]
Short answer: depends.

What are you trying to do? There are no Serverpod modules with ML today, but there are a few ML packages on pub.dev that you could checkout, and see if they meet your needs. Other than this, if you want to use a service that leverages ML, LLM, etc there should be no problem doing so as long as said service has an API. For example building an integration with openai's rest API, or using firebase ML modules.

**Reply by rectrick** [reply]
Unfortunately, tensorflow services are limited, but what if you write intelligence codes in Python and want to create an API for it via serverpod, this will open a very big door to support all artificial intelligence software
  I am able to integrate two projects that I am implementing with Flutter and Serverpod
And if a tool is provided that allows the creation of APIs for AI files and models, I will be able to support this and spread it among the Arab programmers community, as they are interested in integrating these technologies in an integrated manner.

**Reply by Isakdl** [reply]
In theory, it could be done, There's nothing stopping you from creating a Serverpod API that serves as a wrapper for your Python-based ML code. You could make use of Flask or FastAPI to serve your Python models and then communicate with them via Serverpod, as an example. There also seem to exist projects like [DartPy](https://pub.dev/packages/dartpy) to call Python code from Dart, but I have never used this approach myself so I cannot vouch for it.

It is a quite broad topic if we are talking about generic interoperability between Dart and Python. 

But what are you specifically trying to achieve? Do you want a generic ML library to work with and serve the results? If so I think your best bet is building the ML part in existing tools and bridging the data to Serverpod. 

If you are after some specific feature like translation, feature detection in images, word prediction, etc, that might in the future be cool to have as a module for Serverpod.
